cEP-0018.md

```diff
@@ -0,0 +1,266 @@
+        """
+        Loads file's AST into memory
+        """
+        if(lang == 'auto'):
+          ext = get_extension(filename)
+        else:
+          ext = Languages(parse_lang_str(lang)[0])[0]
+        if ext is None:
+            raise FileExtensionNotSupported('Unknown File Type')
+        inputStream = file.readlines()
+        lexer = mapping[ext][0](inputStream)
+        parser = mapping[ext][1](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+
+    @staticmethod
+    def make_walker():
+        """
+        Resolves walker object for appropriate type
+        """
+        if lang == 'auto':
+            return ASTWalker
+        else:
+            return resolve_with_package_info(lang)
+
+    @staticmethod
+    def get_walker(file, filename, lang):
+        """
+        Returns a walker to the required language
+        """
+        ret_class = make_walker(lang)
+        if not ret_class:
+            raise Exception('Un-supported Language')
+        else:
+            return ret_class(file, filename)
+
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+    ...
+```
+
+### Prototype of Py3ASTWalker
+
+These kinds of Walkers will supply a high level walker api
+
+```python
+class BasePyASTWalker(ASTBear):
+    LANGUAGES = { 'Python' }
+    def next_function():
+        """
+        Modify the tree node variable such that it only stops at functions
+        which can be detected by the Python grammar rules
+        """
+
+    def get_imports():
+        """
+        Return list of strings containing import statements from the file
+        """
+
+    def get_methods():
+        """
+        Return list of strings containing all methods from the file
+        """
+    ...
+
+class Py2ASTWalker(BasePyASTWalker):
+    LANGUAGES = { 'Python 2' }
+    def get_print_statements():
+        """
+        Return list of print statements, since print is a keyword in py2
+        """
+    ...
+
+class Py3ASTWalker(BasePyASTWalker):
+    LANGUAGES = { 'Python3' }
+    def get_integer_div_stmts():
+        """
+        Returns a list of integer divide statements, since / and // are
+        different in py3
+        """
+    ...
+```
+
+### Prototype of  `ASTBear` class implementation
+
+```python
+from coantlib import ASTWalker
+from coalib.bears.LocalBear import Bear
+
+class ASTBear(LocalBear):
```

Comment:
This bear should have a class method I feel. This classmethod looks at the LANGUAGES property of a class and returns the appropriate walker.